metaflow/plugins/cards/component_serializer.py

```diff
@@ -16,11 +22,267 @@ def get_card_class(card_type):
     return filtered_cards[0]
 
 
+def _component_is_valid(component):
+    """
+    Validates if the component is of the correct class.
+    """
+    if not issubclass(type(component), MetaflowCardComponent):
+        return False
+    return True
+
+
+def warning_message(message, logger=None, ts=False):
+    msg = "[@card WARNING] %s" % message
+    if logger:
+        logger(msg, timestamp=ts, bad=True)
+
+
 class WarningComponent(ErrorComponent):
     def __init__(self, warning_message):
         super().__init__("@card WARNING", warning_message)
 
 
+class ComponentStore:
+    """
+    The `ComponentStore` class handles the in-memory storage of the components for a single card.
+    This class has combination of a array/dictionary like interface to access the components.
+
+    It exposes the `append` /`extend` methods like an array to add components.
+    It also exposes the `__getitem__`/`__setitem__` methods like a dictionary to access the components by thier Ids.
+
+    The reason this has dual behavior is because components cannot be stored entirely as a map/dictionary because
```

Comment:
That's great! It will certainly remove all the complexity of managing/updating two objects (a dict and list). We can keep the class as is but make its internal implementation much much simpler!

Reply:
This hasn't changed yet right?

Reply:
This hasn't changed; I think changing this will result in breaking functionality for anything < py37; Currently cards can work < py37

Reply:
I think we are fine with new functionality not working for such old versions of python. This will actually work in python 3.6+ so it's really just 3.5. Up to you but it would simplify things a bit :) .